tools/db_bench_tool.cc

```diff
@@ -5028,7 +5044,148 @@ class Benchmark {
+    int64_t prefix_access;
+    int64_t prefix_keys;
+  };
+
+  // From our observations, the prefix hotness (key-range hotness) follows
+  // the two-term-exponential distribution: f(x) = a*exp(b*x) + c*exp(d*x).
+  // However, we cannot directly use the inverse function to decide a
+  // key-range from a random distribution. To achieve it, we create a list of
+  // PrefixUnit, each PrefixUnit occupies a range of integers whose size is
+  // decided based on the hotness of the key-range. When a random value is
+  // generated based on uniform distribution, we map it to the PrefixUnit Vec
+  // and one PrefixUnit is selected. The probability of one PrefixUnit being
+  // selected is the same as the hotness of this PrefixUnit. After that, the
+  // key can be randomly allocated to the key-range of this PrefixUnit, or we
+  // can based on the power distribution (y=ax^b) to generate the offset of
+  // the key in the selected key-range. In this way, we generate the keyID
+  // based on the hotness of the prefix and also the key hotness distribution.
+  class GenerateTwoTermExpKeys {
+   public:
+    int64_t prefix_rand_max_;
+    int64_t prefix_size_;
+    int64_t prefix_num_;
+    bool initiated_;
+    std::vector<PrefixUnit> prefix_set_;
+
+    GenerateTwoTermExpKeys() {
+      prefix_rand_max_ = FLAGS_num;
+      initiated_ = false;
+    }
+
+    ~GenerateTwoTermExpKeys() {}
+
+    // Initiate the PrefixUnit vector and calculate the size of each
+    // PrefixUnit.
+    Status InitiateExpDistribution(int64_t total_keys, double prefix_a,
+                                   double prefix_b, double prefix_c,
+                                   double prefix_d) {
+      int64_t amplify = 0;
+      int64_t prefix_start = 0;
+      initiated_ = true;
+      if (FLAGS_prefix_num <= 0) {
+        prefix_num_ = 1;
+      } else {
+        prefix_num_ = FLAGS_prefix_num;
+      }
+      prefix_size_ = total_keys / prefix_num_;
+
+      // Calculate the key-range shares size based on the input parameters
+      for (int64_t pfx = prefix_num_; pfx >= 1; pfx--) {
+        double prefix_p = prefix_a * std::exp(prefix_b * pfx) +
+                          prefix_c * std::exp(prefix_d * pfx);
+
+        if (prefix_p < std::pow(10.0, -16.0)) {
+          prefix_p = 0.0;
+        }
+        if (amplify == 0 && prefix_p > 0) {
+          amplify = static_cast<int64_t>(std::floor(1 / prefix_p)) + 1;
+        }
+
+        PrefixUnit p_unit;
+        p_unit.prefix_start = prefix_start;
+        if (0.0 >= prefix_p) {
+          p_unit.prefix_access = 0;
+        } else {
+          p_unit.prefix_access =
+              static_cast<int64_t>(std::floor(amplify * prefix_p));
+        }
+        p_unit.prefix_keys = prefix_size_;
+        prefix_set_.push_back(p_unit);
+        prefix_start += p_unit.prefix_access;
+      }
+      prefix_rand_max_ = prefix_start;
+
+      // Shuffle the key-ranges randomly
+      Random64 rand_loca(prefix_rand_max_);
```

Comment:
I don't get why the random seed is `prefix_rand_max_`

Reply:
@siying the value of prefix_rand_max_ is calculated by the distribution parameters. Therefore, I want to ensure, if user use the same key access distribution, the allocation of different key ranges is the same. Otherwise, user cannot reproduce the benchmarking results.